### PR TITLE
Engine.close should only return when resources are freed

### LINKED ISSUE #25817: testCorruptTranslogTruncationOfReplica fails on checkIndex
example: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+intake/76/

```
FAILURE 0.46s J0 | TruncateTranslogIT.testCorruptTranslogTruncationOfReplica <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: at least one shard failed CheckIndex
   > 	at __randomizedtesting.SeedInfo.seed([3132416A541FE03E:1C1C1DE115498AB5]:0)
   > 	at org.elasticsearch.test.ESTestCase.ensureCheckIndexPassed(ESTestCase.java:430)
````

This has to do with a concurrency issue between shard failures and node shutdown. I have a fix plan and I'll get to this as soon as I can.

### PR DESCRIPTION
**Pull request by bleskes**
Currently Engine.close can return immediately if the engine is already at the process of shutting down (due to a concurrent close call or an engine failure). This is a shame because some of our testing infra wants to do things like checking the index. This commit changes the logic to make sure that all calls to close wait until resources are freed. Failing the engine is still non blocking.

Fixes #25817

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/engine/Engine.java [hunk 1]
```diff
@@ -1300,8 +1302,9 @@ public void release() {
 
     /**
      * Method to close the engine while the write lock is held.
+     * Must decrement {@param closedLatch} when closing work is done
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/index/engine/Engine.java]: Looks like this causes javadocs to be unhappy

**Review by dakrone — APPROVED** [review]
LGTM, left one comment about javadocs

**Review by s1monw — APPROVED** [review]
LGTM

File: core/src/main/java/org/elasticsearch/index/engine/Engine.java (hunk 1 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/Engine.java]: @dakrone aye. I had a commit locally fixing this which I forgot to push. Thanks for noticing.

**Comment by bleskes** [comment]
Thx @dakrone & @s1monw

**Pull request merged**